Feature Add Oracle driver

ytetsuro opened the pull request:
**Description**
I wanted to use CodeIgniter4 in an Oracle environment.
Also, it was determined that support would be provided because ORACLE was described in the document.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

The design of the program to get the insertId of this `PR` is not so good.

Please good solution.

lonnieezell commented:
Thanks so much for submitting this! You rock for that. I won't be able to get to this until after the next RC release, honestly. Besides reviewing this, we need to get Travis setup to be able to test it if possible. I'm honestly not familiar enough with Oracle to know if there's a simple to setup community version. Looks like there's Express Edition but would have to look deeper to see if we could integrate.

ytetsuro commented:
@lonnieezell 
Thanks feedback.

> Looks like there's Express Edition but would have to look deeper to see if we could integrate.

Humm...
it may be difficult.

https://github.com/oracle/docker-images/issues/1156

Is there anything I can do to help with the this review?

yogahilmi commented:
I tried using this driver with Oracle 18C, it works fine.

Thank you @ytetsuro

kenjis commented:
@ytetsuro @lonnieezell 
This repo has GitHub workflows with Oracle image.
https://github.com/yiisoft/db-oracle

michalsn commented:
I was working on tests for this at some point and the main problem I had, was how to set an empty database because by default oracle comes with a lot of additional tables that break our tests. Also, I think we need at least Oracle 12 to support "auto-increment" values.

Maybe this thing with a fresh database is trivial but my DevOps skills are low 😅

jkashwin commented:
Any update on the availability of OCI8 driver support for CI4 ?

javawebmedia commented:
I am really waiting for this Codeigniter 4 Oracle driver. Hopefully, it can be released soon. Thank you

ytetsuro commented:
@kenjis 

> This repo has GitHub workflows with Oracle images.

Very thanks.

Using the same idea, I checked the containers used in version 12c and above and found that `quillbuilduser/oracle-18-xe` was used.

The following repositories use this container.

- https://github.com/getquill/quill
- https://github.com/knex/knex/blob/01cfa98c2cdd9d2f8c91320d4551efdaa1768d21/scripts/docker-compose.yml
- https://github.com/JetBrains/Exposed/blob/b6d00e51062c91c6b23a89de55f900e2e4f24225/buildScripts/docker/docker-compose-oracle.yml

@lonnieezell 
@kenjis 

> we need to get Travis setup to be able to test it if possible

Changed GitHub Actions so that the OCI8 tests work with the above container.
Please review.

kenjis commented on system/Database/OCI8/Builder.php:
```diff
@@ -0,0 +1,293 @@
+use CodeIgniter\Database\Exceptions\DatabaseException;
+
+/**
+ * Builder for MySQLi
```
```suggestion
 * Builder for OCI8
```

kenjis commented on system/Database/OCI8/Forge.php:
```diff
@@ -0,0 +1,310 @@
+namespace CodeIgniter\Database\OCI8;
+
+/**
+ * Forge for MySQLi
```
```suggestion
 * Forge for OCI8
```

kenjis commented on system/Database/OCI8/PreparedQuery.php:
```diff
@@ -0,0 +1,134 @@
+use CodeIgniter\Database\PreparedQueryInterface;
+
+/**
+ * Prepared query for MySQLi
```
```suggestion
 * Prepared query for OCI8
```

kenjis commented on system/Database/OCI8/Result.php:
```diff
@@ -0,0 +1,122 @@
+use stdClass;
+
+/**
+ * Result for OCI
```
```suggestion
 * Result for OCI8
```

kenjis commented on system/Database/OCI8/Utils.php:
```diff
@@ -0,0 +1,38 @@
+use CodeIgniter\Database\Exceptions\DatabaseException;
+
+/**
+ * Utils for MySQLi
```
```suggestion
 * Utils for OCI8
```

kenjis commented:
@ytetsuro @lonnieezell @michalsn @jkashwin @javawebmedia
What version of Oracle Database should this driver support?
12.1+?
12.2+?
18c+?
other?

And we need to add about this driver in
https://codeigniter4.github.io/CodeIgniter4/intro/requirements.html

ytetsuro commented:
@kenjis 

Thanks for the quick feedback.

> What version of Oracle Database should this driver support?

I think `12.1+`.
I used the following version of this driver when I created it.

- 12.1
- 18c

The driver I created this time doesn't seem to use any of the features that were deprecated in 18c, so I think 12.1+ is fine.

https://docs.oracle.com/en/database/oracle/oracle-database/18/upgrd/behavior-change-deprecate-desupport-oracle-12c-rel-2.html#GUID-3BAFD95E-4D00-4F0F-BC80-6064F497F878

kenjis commented on tests/system/Models/PaginateModelTest.php:
```diff
@@ -46,7 +46,7 @@ public function testPaginatePassPerPageParameter(): void
     {
         $this->createModel(ValidModel::class);
-        $this->model->groupBy('id');
+        $this->model->select('id')->groupBy('id');
```
Why do you need `->select('id')`?

ytetsuro replied:
@kenjis 

Thanks, response.

> Why do you need ->select('id')?

The reason is that it is grouped by id.
I think you will get `ERROR 1055 (42000)` with `only_full_group_by` in MySQL, for the same reason.
Similar to the above, Oracle cannot select items that are not group by when using group by.

I think it strange that you can select anything but id.
This is because it is illegal in standard SQL.

https://dev.mysql.com/doc/refman/5.6/en/group-by-handling.html

> For example, this query is illegal in standard SQL because the nonaggregated name column in the select list does not appear in the GROUP BY:

kenjis commented on tests/system/Database/Live/GroupTest.php:
```diff
@@ -43,7 +43,7 @@ public function testHavingBy()
             ->select('name')
             ->groupBy('name')
-            ->having('SUM(id) > 2')
+            ->having('SUM(' . $this->db->protectIdentifiers('id') . ') > 2')
```
I want to remove `$this->db->protectIdentifiers()`, because it is Query Builder.
But how?

ytetsuro replied:
@kenjis 

Thanks, response.

You're right.

> But how?

Hmmm…

If you are using the OCI8 driver, what do you think about doing an ALTER TABLE to rename the id column to an uppercase column name?

This method does not essentially solve the QueryBuiler problem.
However, it may be a good solution in that it does not include this problem in the scope of this PR.

It looks just like this.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/commits/729acdc7df399432bd2f5398c302b367cf019553#diff-c2d356f0d870b5a4a5bf49fad5e365c8ec6b3ee9b0f5a6bcb1f879801cd372d9L206-L215

kenjis replied:
How about converting to uppercase when creating tables?

kenjis replied:
No. Query Builder basically quotes identifier like `"id"`
when `$builder->where('id', 3);`

ytetsuro replied:
@kenjis 

Thanks, response.

> No. Query Builder basically quotes identifiers like "id".

No. can’t escape identifier in parenthesis.

https://github.com/codeigniter4/CodeIgniter4/blob/4649204306276af6c06bfbd229fb1895fb5b0e4e/system/Database/BaseConnection.php#L970-L979

> How about converting to uppercase when creating tables?
> If you are using the OCI8 driver, what do you think about doing an ALTER TABLE to rename the id column to an uppercase column name?

I don't think I can solve either of these problems.
Oracle can handle lowercase letters as long as they are escaped.
However, if they are not escaped, they are treated as uppercase.
In other words, it will fail in the following cases.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/files#diff-4d07390c9284f69a1dca0fd6dd95b0ba949997b35d1266828f0619bef65c7c22R58-R59

In the first having, the id is escaped and treated as lowercase.

The implementation I have in mind now is the following.

1. parse AST from SQL using a library like `PhpMyAdmin\SqlParser`.
2. escape identifiers from the AST.

In this case, we have the following concerns
I don't think it is desirable to run a heavy process many times for issuing a single query, what do you think?

kenjis replied:
This is a issue in the Query Builder. There is no complete SQL statement to parse.
The current implementation of `protectIdentifiers()` may receive the string like a sub query `(SELECT MAX(advance_amount) FROM "orders" WHERE "id" > 2`, which does not have `)` at the end.

It seems to me that it is better to keep the specifications simple, maintain speed, and not do anything too complicated.

If we accept the comment as the specification of the Query Builder:
> If a parenthesis is found we know that we do not need to escape the data or add a prefix. 
> Added exception for single quotes as well, we don't want to alter literal strings.

The Query Builder never escapes items which contain `(`, `)`, and `'`.

In this case, it is recommended to write code like `->having('SUM("id") > 2')`.
No, `->having('SUM("id") >',  2)` is better.

ytetsuro replied:
I think you are right.
However, MySQL uses the escape character "`".
Then my code will look like the following.

```php
$result = $this->db->table('job')
            ->select('name')
            ->groupBy('name')
            ->having('SUM(' . $this->db->protectIdentifiers('id') . ') >',  2);
```

In short, I wanted to erase `$this->db->protectIdentifiers()` the same way you did.
However, I can't think of a way to get rid of it.😭😭😭

kenjis replied:
Okay. We can't solve this problem without making parser for parts of SQL (like `'SUM(id) > 2'`), 
and quotes identifiers automatically.

1. Use Quoted Identifiers
- When we create tables with *Forge* in a normal way, identifiers will be quoted.
- In this case, on Oracle, we always need to use quoted identifiers like `"name"`. 
- But MySQL uses `` ` `` to quote identifiers like `` `name` ``.

2. Use Unquoted Identifiers
- When we create tables with unquoted identifiers, Oracle considers them to be uppercase.
- Query Builder quotes identifiers in most cases.
- If we use a lowercase identifier like `name`, QB converts it to `"name"`, and there is no such identifier.

---

Oracle is a special case. I don't show code like `->having('SUM(' . $this->db->protectIdentifiers('id') . ') >',  2)`
only for Oracle.

How about using if statements to divide cases?

```php
    public function testHavingBy()
    {
        if ($this->db->DBDriver === 'OCI8') {
            $result = $this->db->table('job')
                ->select('name')
                ->groupBy('name')
                ->having('SUM("id") > 2')
                ->get()
                ->getResultArray();
        } else {
            $result = $this->db->table('job')
                ->select('name')
                ->groupBy('name')
                ->having('SUM(id) > 2')
                ->get()
                ->getResultArray();
        }

        $this->assertCount(2, $result);
    }
```

ytetsuro replied:
@kenjis 

> How about using if statements to divide cases?

Very thanks, response.
I agree with this.
I'll modify it this way.

> only for Oracle.

I did not have enough knowledge about this.
I am sorry.

To be precise, it seems to be a specification of ANSI-SQL, not Oracle.
That is, it is ANSI-SQL compliant, and a similar implementation may be needed when implementing `Firebird`, which CodeIgniter will support.

kenjis replied:
It seems you are correct. It comes from ANSI-SQL.

https://stackoverflow.com/questions/153944/is-sql-syntax-case-sensitive/19933159#19933159

kenjis commented on system/Database/BaseBuilder.php:
```diff
@@ -1358,9 +1359,9 @@ public function set($key, $value = '', ?bool $escape = null)
                 $bind = $this->setBind($k, $v, $escape);
 
-                $this->QBSet[$this->db->protectIdentifiers($k, false, $escape)] = ":{$bind}:";
+                $this->QBSet[$this->db->protectIdentifiers($k, false)] = ":{$bind}:";
```
This change changes the behavior.
Even if `$escape` is false, `protectIdentifiers()` escapes the identifier.
https://github.com/codeigniter4/CodeIgniter4/blob/c05f5a6dae2b08e9ebf82d5f7888dd8fcbff5813/system/Database/BaseConnection.php#L954-L958

Users can't stop escaping.

ytetsuro replied:
I decided that there was no case where the column name in the set clause did not need to be escaped, so I turned off the unescaping process.

Are there any use cases where the column names in the set clause do not need escaping?

kenjis replied:
Good point. But it seems it has nothing to do with this Oracle driver.
Can you separate the changes that are not specific to the Oracle driver as another PR?
This PR is too big and takes too much time to review.

ytetsuro replied:
Sorry to make your review so difficult.
I will split it up.💪

ytetsuro replied:
This change was merged into CI, so I rebased it.

kenjis commented on system/Database/BaseBuilder.php:
```diff
@@ -1358,9 +1359,9 @@ public function set($key, $value = '', ?bool $escape = null)
+                $this->QBSet[$this->db->protectIdentifiers($k, false)] = ":{$bind}:";
             } else {
-                $this->QBSet[$this->db->protectIdentifiers($k, false, $escape)] = $v;
+                $this->QBSet[$this->db->protectIdentifiers($k, false)] = $v;
```
The same as above.

ytetsuro replied:
This change was merged into CI, so I rebased it.

kenjis commented on tests/system/Database/Live/PreparedQueryTest.php:
```diff
@@ -114,8 +123,8 @@ public function testExecuteRunsQueryAndReturnsManualResultObject()
-        $query->execute('foo', 'foo@example.com', '');
-        $query->execute('bar', 'bar@example.com', '');
+        $query->execute('foo', 'foo@example.com', 'US');
+        $query->execute('bar', 'bar@example.com', 'GB');
```
Why do you need to change the empty strings?

ytetsuro replied:
Thanks, good comment.

This is because Oracle treats an empty string as null and the value of the country column is NotNull, so the query will fail.
oci8 also treats empty strings as null.

https://github.com/php/php-src/blob/master/ext/oci8/tests/bind_sqltafc.phpt#L149-L175

kenjis replied:
You're correct.
Oracle treats an empty string as null.

kenjis commented on system/Database/Forge.php:
```diff
@@ -466,6 +466,8 @@ public function createTable(string $table, bool $ifNotExists = false, array $att
                 return false;
             }
+
+            return true;
```
Is this a required change only for Oracle drivers?

ytetsuro replied:
You're right!
I created a PR.
However, the only pattern where the `createTableIfStr` property is false is for SQLite less than 3.3, except for OCI8.
Therefore, it may be easier to maintain if you fix it in this PR as the test code is a little difficult to convey the intention as shown below.
Maybe I'm just a bad test code.

https://github.com/codeigniter4/CodeIgniter4/blob/f564f6b7eca2504ac894bbaf7c3b6d18ce9b7b85/tests/system/Database/Forge/CreateTableTest.php#L23-L40

kenjis replied:
#5133 was merged. This change is no longer necessary.

ytetsuro replied:
This change was merged into CI, so I rebased it.

kenjis commented on tests/system/Database/Live/GroupTest.php:
```diff
@@ -40,10 +40,13 @@ public function testGroupBy()
             ->select('name')
             ->groupBy('name')
-            ->having('SUM(id) > 2')
+            ->having($sumIdColumn . ' >', 2)
```
I think it is not good practice that using variable in the key of *having* or *where* method of QB.

ytetsuro replied:
Thanks, response.

Fixed it.

Sure!
Looking at this code makes me want to look for SQL Injection!😂

ytetsuro commented on system/Database/BaseBuilder.php:
```diff
@@ -1003,7 +1003,7 @@ protected function _like($field, string $match = '', string $type = 'AND ', stri
             }
 
-            $likeStatement = $this->_like_statement($prefix, $k, $not, $bind, $insensitiveSearch);
+            $likeStatement = $this->_like_statement($prefix, $this->db->protectIdentifiers($k, false, $escape), $not, $bind, $insensitiveSearch);
```
splited #5170

ytetsuro commented on system/Database/BaseBuilder.php:
```diff
@@ -1260,6 +1260,7 @@ public function orderBy(string $orderBy, string $direction = '', ?bool $escape =
         if ($direction === 'RANDOM') {
             $direction = '';
             $orderBy   = ctype_digit($orderBy) ? sprintf($this->randomKeyword[1], $orderBy) : $this->randomKeyword[0];
+            $escape    = false;
```
@kenjis 

Should we split the code here?

The intention of this fix is because escaping is not necessary for RANDOM.

In ORACLE, random numbers are `DBMS_RANDOM.RANDOM`. That is, it refers to the `RANDOM` column of the `DBMS_RANDOM` table.
In MySQL, it is `RAND()`, it is function.

The third argument `$escape` of orderBy is null if it is omitted.
It depends on the `protectIdentifiers` property of Connection to toggle whether escaping is performed or not.
Also, the escaping process is to add the prefix to the table name.

In other words, in the case of ORACLE, it will add the prefix to  `DBMS_RANDOM.RANDOM` and execute a query for a non-existent table, which will fail.

It is not a good idea for users to specify the value of the third argument escape with the driver in mind.

Therefore, I am making this fix because I thought it would be better to not escape for RANDOM and add the escaped column name to the `randomKeyword` property of each driver.

This is not limited to Oracle, but in the Database supported by CodeIgniter4 at this time, RANDOM is all function calls and cannot be reproduced.

kenjis replied:
The more you split this PR, the more likely it is to be merged.
I'm not sure, how long it will take it to be merged.
Because this PR is too big for anyone to fully review.

ytetsuro replied:
@kenjis 

I understand.
Thank you.
I will do my best to remove all diffs from this PR except `system/Database/OCI8`.💪

ytetsuro replied:
splited #5179

ytetsuro commented on system/Database/BaseConnection.php:
```diff
@@ -1269,8 +1269,12 @@ protected function _escapeString(string $str): string
     {
-        $driver = strtolower($this->DBDriver);
-        $driver = ($driver === 'postgre' ? 'pg' : $driver) . '_';
+        $driver = strtolower($this->DBDriver) . '_';
+        if ($this->DBDriver === 'postgre') {
+            $driver = 'pg_';
+        } elseif ($this->DBDriver === 'oci8') {
+            $driver = 'oci_';
+        }
```
This is the reason why the coverages are treated as being lowered.
What should I do?

It may be possible to fudge the coverage by doing the following, but it is a tricky code.

```php
$driver = strtolower($this->DBDriver);
$driver = (['postgre' => 'pg', 'oci8' => 'oci'][$driver] ?? $driver) . '_';
```

kenjis replied:
Something like this?
```diff
@@ -1269,12 +1273,7 @@ abstract class BaseConnection implements ConnectionInterface
      */
     public function callFunction(string $functionName, ...$params): bool
     {
-        $driver = strtolower($this->DBDriver) . '_';
-        if ($this->DBDriver === 'postgre') {
-            $driver = 'pg_';
-        } elseif ($this->DBDriver === 'oci8') {
-            $driver = 'oci_';
-        }
+        $driver = $this->getDriverFunctionPrefix();
 
         if (strpos($driver, $functionName) === false) {
             $functionName = $driver . $functionName;
@@ -1291,6 +1290,11 @@ abstract class BaseConnection implements ConnectionInterface
         return $functionName(...$params);
     }
 
+    protected function getDriverFunctionPrefix()
+    {
+        return strtolower($this->DBDriver) . '_';
+    }
+
     //--------------------------------------------------------------------
     // META Methods
     //--------------------------------------------------------------------
```

ytetsuro replied:
@kenjis 

Thanks, response.
I think it's really great.
You're a god.

ytetsuro replied:
splited #5178

kenjis commented:
@ytetsuro 
> The design of the program to get the insertId of this PR is not so good.

What's wrong with it?

ytetsuro commented:
@kenjis 

> What's wrong with it? 

My approach is to allow the last insert id to be retrieved only when a row is added with QueryBuilder::insert().

kenjis commented:
> For developers, I think it would be better if the API asked for the table name and explored the SEQUENCE internally, what do you think?

I think passing the table name is better for developers.

> My concern is that insertID() is defined as abstract in the parent class.

If we add a parameter, it is a breaking change.
https://3v4l.org/JTMSP
So we can't include the change in the next release.

We will add that feature in the future and I recommend that it not be included in this PR.

Incidentally, CI3 has the three type of `insert_id()`:
```
public function insert_id()
public function insert_id($generator_name, $inc_by = 0) ... ibase
public function insert_id($name = NULL) ... pdo
```

kenjis commented:
> Would it still be a disruptive change if it does not affect existing users by using the default argument feature as follows?

It is still a breaking change. If a user extends BaseConnection, it breaks.

https://3v4l.org/lnu0u

ytetsuro commented:
@kenjis 

Thanks, good response.

> If a SQL statement is INSERT, we get the table name. To do so, we need to parse the statement.

Good!
If it's an INSERT, the subquery won't be the target, so it'll be easy to parse the table name!

But, I have discovered a new problem.😭
The way we are thinking now, we will get an unintended id.
SEQUENCE is not unique for a connection.
This means that if you and I execute a write operation at the same time, and I get the last id, my process may get the id that you wrote.🤮

MySQL's last_insert_id is also set on a per-connection basis.

> The ID that was generated is maintained in the server on a per-connection basis.

https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id

With this in mind, at the moment I can't think of a way to get the last id with the query() method.😭😭

ytetsuro commented on tests/system/Database/Live/OCI8/LastInsertIDTest.php:
```diff
@@ -0,0 +1,77 @@
+-- INSERT INTO "db_misc" ("key", "value") VALUES ('key', 'value')
+/* INSERT INTO "db_misc" ("key", "value") VALUES ('key', 'value') */
+INSERT /* INTO "db_misc" */ INTO "db_job"  ("name", "description") VALUES (' INTO "abc"', ?)
+SQL;
```
@kenjis 

Thanks, good response.

Fixed it!

ytetsuro commented on system/Database/OCI8/Connection.php:
```diff
@@ -0,0 +1,730 @@
+     * @return string
+     */
+    public function parseInsertTableName(string $sql): string {
+        $commentStrippedSql = preg_replace(['/\/\*(.|\n)*?\*\//m', '/^-- .*/'], '', $sql);
```
Thanks, response.

I fixed it by referring to the attached link.

kenjis commented:
@ytetsuro Thanks!

It seems all the problems except the [bug on storedProcedure()](https://forum.codeigniter.com/thread-76381-post-390991.html#pid390991) are solved.

kenjis commented:
```
There were 104 errors:

1) CodeIgniter\Database\Live\OCI8\CallStoredProcedureTest::testCallPackageProcedure
ErrorException: oci_execute(): ORA-06550: line 1, column 7:
PLS-00201: identifier 'CALCULATOR.PLUS' must be declared
ORA-06550: line 1, column 7:
PL/SQL: Statement ignored

/home/runner/work/CodeIgniter4/CodeIgniter4/system/Database/OCI8/Connection.php:201
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Database/BaseConnection.php:670
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Database/BaseConnection.php:598
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Database/OCI8/Connection.php:520
/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Database/Live/OCI8/CallStoredProcedureTest.php:51
```
https://github.com/codeigniter4/CodeIgniter4/runs/4568866943?check_suite_focus=true

ytetsuro commented:
@kenjis 

Hi.
Very thanks response and long-time review.

> It seems all the problems except the bug on storedProcedure() are solved.

Fixed it.

Please help me.
I want to create stored procedure in migration.
But, happen following error when using oci_parse.
Is it possible to create a stored procedure using the oci8 extension?

`ErrorException: oci_execute(): OCI_SUCCESS_WITH_INFO: ORA-24344: success with compilation error in`

The documentation for oci_parse says to create it in advance.

https://www.php.net/manual/en/function.oci-parse.php#refsect1-function.oci-parse-examples

In the documentation of oci_bind_array_by_name, there is a sample to create a stored procedure, but this sample fails.
The error in case of failure is `ORA-24344`, which is the same error I saw this time.

https://www.php.net/manual/en/function.oci-bind-array-by-name.php#refsect1-function.oci-bind-array-by-name-examples

Current commit, the stored procedure is created in the [workflow of github actions](https://github.com/codeigniter4/CodeIgniter4/pull/2487/files#diff-117de92b2c602fd6341909de2b67f7d0a0cb47a8193c7573c1f6752c38492f68R121).
However, it is not a very satisfactory fix.
Do you know how to create a stored procedure using oci8?

kenjis commented:
Did you see https://www.oracle.com/technetwork/database/database-technologies/php/201212-ug-php-oracle-1884760.pdf ?

*PL/SQL Success With Information Warnings* (P.210 in the PDF) shows code to check for informational errors and warning.

ytetsuro commented:
@kenjis 

Thank you.
I hadn't seen that.

When I checked it, there was a bug in the Connection class that removed the semicolon.
Fixed it.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/commits/c3d82cb59234f07445c7671a07742d33740743cd#diff-dcb48cf393b982829be79ab49001306c5f06420ded11fe7c462462922ec92191L192-L195

This fix requires the removal of the semicolon for code outside of the OCI8 directory.
This is a fix only for OCI8.
Do you think this fix should be put out as a separate PR?

https://github.com/codeigniter4/CodeIgniter4/pull/2487/commits/c3d82cb59234f07445c7671a07742d33740743cd#diff-a07677ecb7c1a5d29e0d0a8e72f00b5b77b577508602ada4fb8c5c1e0c7c5916L1045-L1052

kenjis commented:
> Do you think this fix should be put out as a separate PR?

Yes.
I don't get what the problem is at a sight.
Send another PR and explain it, please.

ytetsuro commented:
@kenjis 

Thanks, response.
I made PR.
Please review.

https://github.com/codeigniter4/CodeIgniter4/pull/5513

ytetsuro commented:
> It seems all the problems except the [bug on storedProcedure()](https://forum.codeigniter.com/thread-76381-post-390991.html#pid390991) are solved.

Fixed it.
Please review.

I think this solves the problem of storedProcedure.
I couldn't confirm the problem 1 he was talking about, but I could confirm the problem 2, so I fixed it.
I also added a test to show that problem 1 could not be confirmed.

lonnieezell commented:
@kenjis @paulbalandan  I think we've all reviewed this in the past. Does anyone see anything we need to do in order to get this merged?

kenjis commented:
@lonnieezell I think this is completed. This feature will be included in v4.2.0.
I'm not sure Ci4 branching, so I hold off on merging.
At least I'm going to merge when `develop` will become a branch for 4.2.0.

javawebmedia commented:
> @lonnieezell I think this is completed. This feature will be included in v4.2.0. I'm not sure Ci4 branching, so I hold off on merging. At least I'm going to merge when `develop` will become a branch for 4.2.0.

waiting for the new features

kenjis commented:
@ytetsuro CI 4.2 requires PHP 7.4 or later, and coding style has been changed for 7.4.
Can you rebase and run rector to fix the GitHub Action errors?

ytetsuro commented:
@kenjis 

Very thanks long time review.
Of cource, i can!

lonnieezell commented:
@ytetsuro thanks for sticking with us on this one! Once you get those changes taken care of we'll be able to finally merge!

javawebmedia commented:
Thank you so much. Hopefully it is going to be release soon.

kenjis commented:
@ytetsuro This is one of the new big features in v4.2.0.
Can you add a short description to the changelog?
https://github.com/codeigniter4/CodeIgniter4/blob/develop/user_guide_src/source/changelogs/v4.2.0.rst#enhancements

kenjis commented on user_guide_src/source/changelogs/v4.2.0.rst:
```diff
@@ -23,6 +23,8 @@ Enhancements
     - See :ref:`content-security-policy` for details.
 - New View Decorators allow modifying the generated HTML prior to caching. 
+- Added new OCI8 driver for database.
+    - You to access Oracle Database. They support SQL and PL/SQL statements.
```
```suggestion
    - It can access Oracle Database and supports SQL and PL/SQL statements.
```

ytetsuro commented:
@kenjis 

Thanks for, feedback!

> Can you rebase and run rector to fix the GitHub Action errors?

Fixed it!💪

> Can you add a short description to the changelog?

Added it!
In addition, Added a document to the query method about the presence of semicolons for OCI8.
Please review.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/commits/4698fc6b1b2a4ab36e2dff3910d060407d51f663

kenjis commented on user_guide_src/source/database/queries.rst:
```diff
@@ -27,6 +27,9 @@ this::
 
     $query = $db->query('YOUR QUERY HERE');
 
+.. note:: If you are using OCI8 Driver, SQL statements should not end with a semi-colon (`;`).
+    PL/SQL statements should end with a semi-colon (`;`).
```
```suggestion
.. note:: If you are using OCI8 Driver, SQL statements should not end with a semi-colon (``;``).
    PL/SQL statements should end with a semi-colon (``;``).
```

kenjis reviewed (changes requested):
I added an inline comment.
And can you resolve the conflict?

I would like to merge as soon as possible.

ytetsuro commented:
@kenjis 

Thanks for, review!

I Fixed it!!💪

I think should add storedProcedure method in the document.
But where to write it I do not know.😭

https://forum.codeigniter.com/thread-76381-post-391902.html#pid391902

Another question is that the storedProcedure method seems to have been removed in CI3.🤔
If the reason is reasonable, I would like to remove the storedProcedure method in this PR as well.

https://github.com/bcit-ci/CodeIgniter/issues/3981#issuecomment-1006056197

What do you think?

kenjis commented:
@ytetsuro Please fix coding style.

kenjis commented:
I don't know why `stored_procedure()` was removed in v3.
Since v3 and v4 are developed separately, there is no need to make them the same.

I think deleting it will cause problems for users who are already using it.

ytetsuro commented:
> Please fix coding style.

Fixed following!💪

https://github.com/codeigniter4/CodeIgniter4/pull/5581

kenjis commented on system/Database/OCI8/Connection.php:
```diff
@@ -0,0 +1,720 @@
+     *
+     * @return mixed
+     *
+     * params array keys
+     *
+     * KEY      OPTIONAL  NOTES
+     * name     no        the name of the parameter should be in :<param_name> format
+     * value    no        the value of the parameter.  If this is an OUT or IN OUT parameter,
+     *                    this should be a reference to a variable
+     * type     yes       the type of the parameter
+     * length   yes       the max size of the parameter
```
Since this description is about parameters, please modify it to the `@param` format.

ytetsuro replied:
Thanks, the feedback.

Fixed it!.

kenjis commented on system/Database/OCI8/Connection.php:
```diff
@@ -0,0 +1,720 @@
+        // Build the query string
+        $sql = 'BEGIN ' . $procedureName . '(';
+
+        $haveCursor = false;
```
`$haveCursor` is not used. Is it okay?

ytetsuro replied:
No problem.

Added test.🤟

ytetsuro commented:
> I think deleting it will cause problems for users who are already using it.

Thanks for, feedback!!

I sure.

kenjis commented on system/Database/OCI8/Connection.php:
```diff
@@ -0,0 +1,720 @@
+     *
+     * @param string $procedureName procedure name to execute
+     *
+     * @return mixed
```
Could you make it more specific types?

ytetsuro replied:
Thanks, for feedback.

Fixed it!.

kenjis commented:
> Random ordering is not currently supported in Oracle and will default to ASC instead.
https://codeigniter4.github.io/CodeIgniter4/database/query_builder.html?highlight=oracle

It this correct?

ytetsuro commented:
@kenjis 

> I don't know why stored_procedure() was removed in v3.
> Since v3 and v4 are developed separately, there is no need to make them the same.
> I think deleting it will cause problems for users who are already using it.

I am tempted to delete storedProcedure and getCursor after seeing this reply and narfbg's reply.
https://github.com/bcit-ci/CodeIgniter/issues/3981#issuecomment-1013785678

I understand that CodeIgniter3 and CodeIgniter4 are different, but as narfbg said, Database Driver abstracts Database operations, and this is the same for CodeIgniter4.

The return value of db_connect is also BaseConnection, and the user's IDE cannot determine that those methods exist.

I also agree that it would be annoying to users who are already using it.

To that end, here are the solutions I'm currently considering.
Which do you think is better?

1. create a file like Helpers/oci8_helper to realize storedProcedure in a different module from Database abstraction.
1. stop using original prepared statements when using OCI, and use a class that uses oci_bind_by_name to create queries.
1. Or if you have a better solution, please let me know.

ytetsuro commented:
> It this correct?

Thanks for feedback.

Removed it.

kenjis commented:
> Database Driver abstracts Database operations, and this is the same for CodeIgniter4.

I'm not sure CI4 aims to Database Driver abstraction.
If we aim to it, we cannot use any Database specific functionality,
and cannot add any Database Driver specific functionality to a specific driver.
We should remove `callFunction()`, because it does not work if a user changes the database.

Personally I think abstraction is Query Builder (and CodeIgniter\Model) and Forge.
And Stored Procedure is not so rare functionality, but not all databases have it.
It might be possible that BaseConnection has `storedProcedure()`.

> 2. stop using original prepared statements when using OCI, and use a class that uses oci_bind_by_name to create queries.

I don't get what you mean. What is *original prepared statements* ?

ytetsuro commented:
@kenjis 

Thanks for feedback.

I'm learning a lot. Thank you.

> I'm not sure CI4 aims to Database Driver abstraction.
> If we aim to it, we cannot use any Database specific functionality,
> and cannot add any Database Driver specific functionality to a specific driver.

Indeed, I don't think we should add it!

 I thought that inheriting from BaseConnection meant that I was abstracting the Database connection.
I don't think we should do that even as a Liskov substitution principle.
If it is to be a subclass of BaseConnection.

However, the idea that you should use PDO if you want abstraction may also exist...

> We should remove callFunction(), because it does not work if a user changes the database.
> Personally I think abstraction is Query Builder (and CodeIgniter\Model) and Forge.

I don't know why there is a `callFunction` that can only return boolean.
However, as you said, abstraction may not be the goal when there are callFunction and query methods.

Thank you. I learned a lot and understand.

> And Stored Procedure is not so rare functionality, but not all databases have it.
> It might be possible that BaseConnection has storedProcedure().

It's true that storedProcedure is not an ORACLE-specific feature.
MySQL can call storedProcedure in a CALL statement via the query method.
In OCI, I think we need to prepare a separate storedProcedure method because we need to allocate variables of reference and resource types.
I think because the binding process for prepared statements is a proprietary implementation.
This is because references and resource types can only be bound using functions such as oci_bind_by_name.

From the above, I feel that it would be better to implement it in 2.
What do you think?

> I don't get what you mean. What are original prepared statements?

I'm sorry.
My explanation was not good.
I meant to stop the original binding implementation for prepared statement.
Specifically, we are considering whether it would be possible to design the binding process in the Query class so that the PreparedQuery class is used for binding.
By doing so, we believe that users will be able to execute storedProcedure in the query method, just like in MySQL.

kenjis commented:
```
$ composer cs-fix
$ vendor/bin/rector process file1 file2 ...
```

kenjis commented:
I would like to release Oracle driver with functionality that can upgrade CI 3.1 code to 4.x as soon as possible.
If we can execute stored procedures by the `query()` method, Does it provide value to the Oracle users?

ytetsuro commented:
> If we can execute stored procedures by the query() method, Does it provide value to the Oracle users?

No.

I can't offer much value in being able to run storedProcedure in the query method.
However, I think I can provide value in stopping proprietary bindings.
For example, it will be able to store binary data.
Being able to execute storedProcedure from the query method is one such example.

https://www.php.net/manual/en/function.oci-bind-by-name.php#example-1789

kenjis commented:
@ytetsuro I don't fully understand your 2. plan, but it seems a good way to go.

But this PR is too long living and I really want to finish it.
Refactoring or improvements can be added later.

Can you fix the coding style error?

kenjis commented on system/Database/OCI8/PreparedQuery.php:
```diff
@@ -0,0 +1,134 @@
+        $count = 0;
+
+        return preg_replace_callback('/\?/', static function ($matches) use (&$count) {
+            return ':' . ($count++);
```
The PHPDoc was fixed, starting with `:0`.

kenjis reviewed (approved):
LGTM

ytetsuro commented:
@kenjis 

Thank you very much for supporting this PR for a very long time.

> I don't fully understand your 2. plan, but it seems a good way to go.

I understood.
I would like to issue an issue with more details on this content when this PR is merged.

Thank you very much!!

kenjis commented:
The one of the oldest, and most commented PR has been merged!
